Tiebout - Apple Push Notification Server App

loganallenc opened the pull request:
Simple gall app for sending push notifications to an urbit user's iPhone when they receive a Hall message from a circle that has been subscribed to these notifications.

vvisigoth commented:
@loganallenc remember to request review (I just added @ixv and @Fang- )

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+    ?<  ?=(~ t.wir)
+    :_  this
+    =/  doc/dock  [our.bol %hall]
+    [ost.bol %peer /our/[i.t.wir] doc /circle/[i.t.wir]/config/grams]~
```
Maybe we want to re-subscribe from the last-heard message going forward? I think the logic below ensures we don't send notifications for whatever old junk we may get here, but still.

Pretty sure this just gives you the last `100` messages, or all messages from the last `~d1` or something along those lines. Being more specific allows you to catch up from even further back when needed, and doesn't get you more info than you need.

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+    ::  %gram: send notification if envelope is lower than read number
+    ::
+      %gram
+    =/  red  (~(get by circles.tie.sta) nom)
```
Indent this stuff (in this `?+`) like you did in earlier code. Current indentation is a tad hard to read.

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+      :_  this(circles.tie.sta (~(put by circles.tie.sta) nom u.red))
+      (message-to-notification u.red nev.rum.rum)
+    :_  this
+    (message-to-notification u.red nev.rum.rum)
```
Wait, huh. The comment says "send notification if ... lower than read number", but it seems to send a notification either way, only updating state in one of the cases? This seems wrong.

(And if it's right, at least do `:-  (msg-to-not ...)  ?:  ...`.)

(There's also questions here about what we want to happen in the "message was edited case", but we probably want to do nothing? Editing isn't really supported anywhere anyway.)

loganallenc replied:
Yes, msg-to-not contains the conditional that checks whether to send a notification. Renamed for clarity.

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+      %config
+    ?+  -.dif.rum.rum
+      [~ this]
+          %read
```
Indents seem weird here. I think `%read` and onward is too deep? Also separate these with a `::` line for absolute clarity.

loganallenc replied:
done

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+::
+++  message-to-notification
+  |=  [red=@ud env=envelope:hall]
+  ^-  (list move)
```
This seems to always produce a list of either zero moves, or one move. Sounds like a `(unit move)` to me.

loganallenc replied:
I like the code style of sending back a list much more. I'd need to add another layer of conditional checks in diff-hall-rumor otherwise.

Fang- replied:
Yeah, I guess `(list move)` is still fair game here. The contract between "notification" and "moves" is a bit weird, so maybe comment, but approved either way.

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+    ~
+  =/  pay  %-  my  :~
+    alert+s+'New message from {(cite:title aut.gam.env)}'
+  ==
```
This could do with a type for `pay`. `my` isn't commonly used, `~(gas in *(map @t json))` is more verbose/clear on what it does (and type-checks early!). For this case, you could also `~(put by *(map @t json))` since you only have one element.

Or just do the dumbest thing and say `[alert+s+'etc' ~ ~]` and cast that to a `map`. (^:

(The most aggressive thing to do here, of course, is to tell @pilfer-pandex to put "map/set syntax" on the global hoon wishlist.)

pilfer-pandex replied:
Yes, this is certainly on my radar.

Fang- replied:
Also wait a second, `'cords'` don't support string interpolation! You need to `(crip "tape {etc}")` here.

loganallenc replied:
I don't like gas in map. It's less clear to me than my.

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+  =/  url=purl  (need (de-purl:html furl))
+  =/  jon=json  :-  %o  %-  my  :~
+    aps+o+payload.not
+  ==
```
[o, my payload](https://i0.kym-cdn.com/entries/icons/original/000/023/390/O.png)

Same map building advice from above applies here (and below). But also, there's stdlib functions dedicated to construction json. Look at `++en-json`.

(Also chaining tall-form together on the same line like this is generally bad style.)

loganallenc replied:
Will fix tall-form chaining.

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+  |=  nom=name:hall
+  ^-  (quip move _this)
+  :_  this(circles.tie.sta (~(put by circles.tie.sta) nom 0))
+  [ost.bol %peer /our/[nom] [our.bol %hall] /circle/[nom]/config/grams]~
```
There's a lot of move construction happening in separate arms. There's common themes, like `ost.bol` at the head, or `[our.bol %hall]` being used for the `dock`. You may or may not want to write arms that deal with the common parts of move construction for you.

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+  ?:  (lte num.env red)
+    ~
+  =/  doc/dock  [king.tie.sta dap.bol]
+  [ost.bol %poke /ask-king doc %tiebout-action [%notify not]]~
```
Ah, so kings route your notifications for you? Is this because we don't want any old ship to access our push endpoint? idk about loading it all onto ~dopzod, we may want a dedicated ship for this, but the load probably isn't *that* big?

Fang- commented on mar/tiebout-action.hoon:
```diff
@@ -0,0 +1,40 @@
+    ++  sa            :: string as ta
+      |=  jon=^json
+      ?>  ?=([%s *] jon)
+      (scot %tas p.jon)
```
If you want a `@tas` from the `@t` `p.jon`, you can either just cast it, or `?>  (sane %tas p.jon)` and then cast, or similar. `(scot %tas` just gives you the "rendered" version, which may not be what you want.

Fang- commented on mar/tiebout-action.hoon:
```diff
@@ -0,0 +1,40 @@
+    %-  (hard action:tiebout)
+    =<  (action jon)
+    |%
+    ++  action
```
`|^` exists and might be nicer here.

loganallenc replied:
Nest-failed in a weird way, skipping it.

Fang- replied:
Happy to help you figure it out if you want to, but I won't press you on it.

Fang- commented on mar/tiebout-action.hoon:
```diff
@@ -0,0 +1,40 @@
+  ++  json
+    |=  jon=^json
+    ^-  action:tiebout
+    %-  (hard action:tiebout)
```
Do you *need* to hard this? If you write your json parsing correctly it should just nest.

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+::
+::::  /app/tiebout/hoon
+::
+/?  309
```
Woah, I haven't seen one of these used in forever. I don't actually know how this operates across breaches that pass by this version etc. (Not that I've ever seen zuse version be changed in the first place.)

loganallenc replied:
removing

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+/?  309
+/-  hall
+/+  tiebout
+=,  tiebout
```
I learned recently that you can do `/+  *tiebout` to not put a face on it, so you don't need to `=,` afterwards.

loganallenc replied:
cool!

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+::
+::  state:
+::
+~%  %tiebout  ..^is  ~
```
Why is this here (and in the `/lib`)? Is this gonna be jetted or something? I haven't ever seen this in apps before.

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+      (del-circle +.act)
+    $notify
+      (send-notify +.act)
+  ==
```
Old syntax!

The indentation is backwards here, I'd just single-line each of these cases anyway. They're short enough.

loganallenc replied:
fixed

ixv commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+::  state:
+::
+~%  %tiebout  ..^is  ~
+|_  [bol=bowl:gall sta=state]
```
I disagree with this, having a face for state means makes things clearer. I've heard other people argue both sides of this before. Perhaps this is something that should be resolved by a style guide. As someone who has had to update someone else's =* lark expression, I'm against that solution.

Fang- replied:
Agreed that `=*`ing some lark is bad. I've felt your pain before. (But never in this context, because the lark here practically never changes.)

ime this is a non-issue for top-level app cores though. Have never ever needed a face on app state. Clarity is resolved by "common pattern" and having the state structure in the same file.

Surprised to hear others have argued for adding a face in this context, would be curious to hear from them.

loganallenc replied:
Keeping the face.

loganallenc commented on lib/tiebout.hoon:
```diff
@@ -0,0 +1,58 @@
+/-  hall
+=,  eyre
+::
+~%  %tiebout-lib  ..is  ~
```
no, removing

loganallenc commented on lib/tiebout.hoon:
```diff
@@ -0,0 +1,58 @@
+  ==
+::
++$  state
+  $%  [%0 tie=tiebout-zero]
```
good idea

loganallenc commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+      %our
+    ?<  ?=(~ t.wir)
+    :_  this
+    ~
```
yup

loganallenc commented on lib/tiebout.hoon:
```diff
@@ -0,0 +1,58 @@
+::
+~%  %tiebout-lib  ..is  ~
+|%
++$  move  [bone card]
```
Moved. Thanks!

loganallenc commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+  ^-  (quip move _this)
+  ~&  prize+[wir piz]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
```
fixed

loganallenc commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,253 @@
+      {%our @ @}
+    ?>  ?=(%circle -.piz)
+    =/  nom/name:hall  i.t.wir
+    =/  red/@ud  red.loc.cos.piz
```
Can you include type information in `=*`?

Fang- replied:
Yes! Since #585 you can do `=*  foo=bar  baz`. I actually forget this made it in already, and should've been using it more. Neat!

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,236 @@
+    [~ this]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  i.wir
```
I thought I left a comment about this in the initial review but can't find it, so maybe it got eaten? You already do a saner `?+` in the `++diff-hall-*` cases, you'll want to apply that here and in `++quit` too.

(Also indenting for this particular one is borked.)

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,236 @@
+  =/  not/notification  :+
+    token.sta
+    'com.tlon.urbit-client'
+    pay
```
You're probably fine putting these between `[]` on one line. If you feel that's too long, proper `:+` form would be

```hoon
:+  token.sta
  'com.tlon.urbit-client'
pay
```

Fang- commented on sur/tiebout.hoon:
```diff
@@ -0,0 +1,62 @@
+::
+::::  /lib/tiebout/hoon
```
Pretty sure filepaths in file headers is outdated style.

And as mentioned before, lots of the structures (cards, state) here really belong in the `/app` file instead, since they aren't shared to/used by other applications/libraries interfacing with Tiebout.

loganallenc commented:
Hey man! I removed the lib from the repository, made the indentation change and removed the `?~  i.wir` you've mentioned. I also added the change to the +:  for the notification that you mentioned. This should be ready to go now. Thanks!

Fang- commented on app/tiebout.hoon:
```diff
@@ -0,0 +1,269 @@
+  ::
+  ::  %our
+  ::
+      {%our @ @}
```
While these are technically correct and will successfully match `/our/whatever` aka `[%our %whatever ~]`, it's better to describe a limited-length wire by actually ending with `~` in the type description.
